File: com.android.tools.build/4.0.1/com.android.tools/sdk-common/27.0.1/com/android/ide/common/gradle/model/IdeLibrary.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model;

import static com.android.ide.common.gradle.model.IdeLibraries.computeResolvedCoordinate;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.builder.model.Library;
import java.io.Serializable;
import java.util.Objects;

/** Creates a deep copy of a {@link Library}. */
public abstract class IdeLibrary implements Library, Serializable {
    // Increase the value when adding/removing fields or when changing the serialization/deserialization mechanism.
    private static final long serialVersionUID = 3L;

    @NonNull private final IdeMavenCoordinates myResolvedCoordinates;
    @Nullable private final String myBuildId;
    @Nullable private final String myProject;
    @Nullable private final String myName;
    @Nullable private final Boolean myIsSkipped;
    @Nullable private final Boolean myProvided;
    private final int hashCode;

    // Used for serialization by the IDE.
    IdeLibrary() {
        myResolvedCoordinates = new IdeMavenCoordinates();
        myBuildId = null;
        myProject = null;
        myName = null;
        myIsSkipped = null;
        myProvided = null;

        hashCode = 0;
    }

    protected IdeLibrary(@NonNull Library library, @NonNull ModelCache modelCache) {
        myResolvedCoordinates = computeResolvedCoordinate(library, modelCache);
        myBuildId = IdeModel.copyNewProperty(library::getBuildId, null);
        myProject = IdeModel.copyNewProperty(library::getProject, null);
        myName =
                IdeModel.copyNewProperty(
                        library::getName, null); // Library.getName() was added in 2.2
        myProvided = IdeModel.copyNewProperty(library::isProvided, null);
        myIsSkipped = IdeModel.copyNewProperty(library::isSkipped, null);
        hashCode = calculateHashCode();
    }

    @Override
    @Nullable
    public IdeMavenCoordinates getRequestedCoordinates() {
        throw new UnusedModelMethodException("getRequestedCoordinates");
    }

    @Override
    @NonNull
    public IdeMavenCoordinates getResolvedCoordinates() {
        return myResolvedCoordinates;
    }

    @Nullable
    @Override
    public String getBuildId() {
        return myBuildId;
    }

    @Override
    @Nullable
    public String getProject() {
        return myProject;
    }

    @Override
    @Nullable
    public String getName() {
        return myName;
    }

    @Override
    public boolean isSkipped() {
        if (myIsSkipped != null) {
            return myIsSkipped;
        }
        throw new UnsupportedOperationException("Unsupported method: IdeLibrary.isSkipped()");
    }

    @Override
    public boolean isProvided() {
        if (myProvided != null) {
            return myProvided;
        }
        throw new UnsupportedOperationException("Unsupported method: IdeLibrary.isProvided()");
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof IdeLibrary)) {
            return false;
        }
        IdeLibrary library = (IdeLibrary) o;
        return library.canEqual(this)
                && Objects.equals(myIsSkipped, library.myIsSkipped)
                && Objects.equals(myProvided, library.myProvided)
                && Objects.equals(myResolvedCoordinates, library.myResolvedCoordinates)
                && Objects.equals(myBuildId, library.myBuildId)
                && Objects.equals(myProject, library.myProject)
                && Objects.equals(myName, library.myName);
    }

    public boolean canEqual(Object other) {
        // See: http://www.artima.com/lejava/articles/equality.html
        return other instanceof IdeLibrary;
    }

    @Override
    public int hashCode() {
        return hashCode;
    }

    protected int calculateHashCode() {
        return Objects.hash(
                myResolvedCoordinates, myBuildId, myProject, myName, myProvided, myIsSkipped);
    }

    @Override
    public String toString() {
        return "myResolvedCoordinates="
                + myResolvedCoordinates
                + ", myBuildId='"
                + myBuildId
                + '\''
                + ", myProject='"
                + myProject
                + '\''
                + ", myName='"
                + myName
                + '\''
                + ", myProvided="
                + myProvided;
    }
}

File: com.android.tools.build/4.2.0-alpha08/com.android.tools/sdk-common/27.2.0-alpha08/com/android/ide/common/gradle/model/IdeLibrary.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model

import java.io.File

/**
 * Represent a variant/module/artifact dependency.
 */
interface IdeLibrary {

  enum class LibraryType {
    LIBRARY_ANDROID,
    LIBRARY_JAVA,
    LIBRARY_MODULE
  }

  /**
   * The type of the dependency
   */
  val type: LibraryType

  /**
   * Returns the artifact address in a unique way.
   *
   *
   * This is either a module path for sub-modules (with optional variant name), or a maven
   * coordinate for external dependencies.
   */
  val artifactAddress: String

  /** Returns the artifact location.  */
  val artifact: File

  /**
   * Returns the build id.
   *
   * This is only valid if the [.getProjectPath] is not null. However this can still be
   * null if this is the root project.
   */
  val buildId: String?

  /**
   * Returns the gradle path.
   *
   * Only valid for Android Library where [type] is [LIBRARY_MODULE]
   */
  val projectPath: String?

  /**
   * Returns an optional variant name if the consumed artifact of the library is associated to
   * one.
   *
   * Only valid for Android Library where [type] is [LIBRARY_MODULE]
   */
  val variant: String?

  /**
   * Returns the location of the unzipped bundle folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val folder: File?

  /**
   * Returns the location of the manifest relative to the folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val manifest: String

  /**
   * Returns the location of the jar file to use for compiling and packaging. The path may not point to an existing file.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID].
   */
  val jarFile: String

  /**
   * Returns the location of the jar file to use for compilation. The path may not point to an existing file.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID].
   */
  val compileJarFile: String

  /**
   * Returns the location of the res folder. The file may not point to an existing folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val resFolder: String

  /**
   * Returns the location of the namespaced resources static library (res.apk). Null if the library is not namespaced.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   *
   * TODO(b/109854607): When rewriting dependencies, this should be populated with the
   * rewritten artifact, which will not be in the exploded AAR directory.
   */
  val resStaticLibrary: File?

  /**
   * Returns the location of the assets folder. The file may not point to an existing folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val assetsFolder: String

  /**
   * Returns the list of local Jar files that are included in the dependency.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val localJars: Collection<String>

  /**
   * Returns the location of the jni libraries folder. The file may not point to an existing folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val jniFolder: String

  /**
   * Returns the location of the aidl import folder. The file may not point to an existing folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val aidlFolder: String

  /**
   * Returns the location of the renderscript import folder. The file may not point to an existing folder.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val renderscriptFolder: String

  /**
   * Returns the location of the proguard files. The file may not point to an existing file.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val proguardRules: String

  /**
   * Returns the location of the lint jar. The file may not point to an existing file.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val lintJar: String?

  /**
   * Returns the location of the external annotations zip file (which may not exist).
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val externalAnnotations: String

  /**
   * Returns the location of an optional file that lists the only resources that should be
   * considered public. The file may not point to an existing file.
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val publicResources: String

  /**
   * Returns the location of the text symbol file
   *
   * Only valid for Android Library where [type] is [LIBRARY_ANDROID]
   */
  val symbolFile: String

  /**
   * Returns whether the dependency is on the compile class path but is not on the runtime class
   * path.
   */
  val isProvided: Boolean
}